### PR TITLE
feat(hitl): durable HITL pause/respond — drop 180s timeout, recover across workers

### PR DESCRIPTION
**Pull request by xfgong**
## Summary

Replaces the 180-second HITL timeout with a durable pause/resume mechanism. A paused HITL request (sandbox confirm or `ask_user`) persists to Postgres, the worker releases (no more in-process `asyncio.Future` blocking), and any worker can resume the conversation when the user answers.

Same `run_id` flows through pause→respond→abort so the SSE stream, traces, and frontend cards stay continuous.

Spec + plan + 5 rounds of local-codex review at:
- `docs/dev/specs/2026-06-02-hitl-checkpointed-respond-design.md`
- `docs/dev/plans/2026-06-02-hitl-checkpointed-respond.md`

Cubepi prerequisite: [cubeplexai/cubepi#135](https://github.com/cubeplexai/cubepi/pull/135) (merged); pin bumped to `1686eed` (v3 schema with new `cubepi_threads.run_id` column persisted atomically alongside `pending_request`).

## What changes

**Backend (cubebox)**
- `paused_hitl` run status; `_FORCE_CLAIM_STALE_LUA` guards it from concurrent `start_run` clobbering
- `claim_resume` Lua CAS for single-flight resume (handles paused / stale / TTL-rebuild branches)
- `_run_cubepi_path` swaps `InMemoryChannel` → `CheckpointedChannel(run_id=...)` (atomic pending+run_id write)
- Auto-detach: `_on_event` listener schedules `agent.detach()` on `HitlRequestEvent` so the worker releases immediately
- `_classify_terminal_status` + dangling-pending cleanup (org-policy-changed-mid-pause edge case → synthetic `*_resolved` event with `policy_overridden` decision)
- Extracted `_build_agent_for_conversation` factory; shared by prompt / respond / cancel paths
- New `RunManager.resume_run_with_answer` + `cancel_paused_run`; old `dispatch_ask_user_answer` / `dispatch_hitl_answer` + their Redis-pubsub plumbing deleted
- `start_run` rejects paused_hitl + DB pending non-null (worker-crash crash window)
- Cancel / steer / cancel_steer routes return 409 `paused_hitl` instead of silent `no_active_run`
- Scheduled-task `dispatch` distinguishes paused (`skipped_paused`, no retry) from busy (existing retry policy)
- Tracing: `hitl.run_id` + `hitl.detached` span attrs; `cubepi.metadata.run_id` + `turn_kind=prompt|respond|abort` on the `invoke_agent` span

**Frontend**
- Bootstrap response carries `pending_hitl` (kind-tagged union); cold-start renders the card without SSE replay
- `PendingAsk`/`PendingConfirm` gain `run_id`; live SSE handlers populate from `currentRunId`
- Composer locks when `pendingConfirmMap` non-empty or `pendingAsk` non-null
- Submit handlers map 409 `resume_in_flight` / `stale_answer` / `conversation_moved` and 404 `no_pending` to user-visible state
- `policy_overridden` decision/reason union extension; existing card removal still runs

**Schema**
- New alembic revision `060310ecfd8a_cubepi_v2_to_v3_pending_run_id` — adds `cubepi_threads.run_id TEXT NULL` + bumps `cubepi_schema_version` 2→3 (matches cubepi PR #135's `EXPECTED_SCHEMA_VERSION`)

## Test plan

- [x] Backend unit: 1212 pass / 4 skip / 0 errors
- [x] Backend e2e (excl. real_llm): 476 pass / 5 skip / 19 deselected; 4 failures all pre-existing cross-test pollution (pass in isolation on both main and this branch)
- [x] Frontend: 50 core + 129 web tests all pass; typecheck + lint clean
- [x] mypy strict clean (317 source files); ruff clean
- [x] New e2e: `tests/e2e/test_hitl_pause_resume.py` (10 scenarios — bootstrap pending_hitl for both kinds, long-pause recovery, claim CAS, two-tab race, stale/no-pending error codes, paused steer/cancel routes)
- [ ] Manual UI smoke: pause an `ask_user` → close tab → wait 30 min → reopen → answer → verify continuation (post-merge)

@codex please review

### CODE REVIEW

File: backend/cubebox/streams/run_manager.py [hunk 1]
```diff
@@ -573,8 +676,21 @@ async def start_run(
...
+            if existing and existing.status in ("running", "paused_hitl"):
                 raise RuntimeError(f"Conversation {conversation_id} already has an active run")
+            # Covers the worker-crash window where Redis meta aged out (or was
+            # never written) but the DB-persisted pending HITL request lingers.
+            # Without this guard a new user turn would race the resume path.
+            from cubebox.agents.checkpointer import init_checkpointer
+
+            async with init_checkpointer() as _cp:
+                _db_pending = await _cp.load_pending_request(conversation_id)
+            if _db_pending is not None:
```
[the remark this answers is not in the document]
Comment (xfgong) [reply backend/cubebox/streams/run_manager.py]: Fixed in a4fb509a. Two-part: (1) `_execute_run`'s finally skips `clear_active_run` + `expire_run_data` when `final_status == "paused_hitl"`, so the Redis key stays as the paused lock and the existing T3 conflict-branch DB guard fires correctly on a racing send. (2) Belt-and-braces: moved the DB-pending check BEFORE `create_run` so a TTL-expired lock (long pause beyond `run_event_ttl_seconds`) still rejects new turns — DB is authoritative. Updated `test_start_run_succeeds_when_no_pending` to assert the new up-front guard is awaited once per start.

File: backend/cubebox/streams/run_manager.py [hunk 2]
```diff
@@ -999,927 +1210,1676 @@ async def _run_cubepi_path(
...
+
+                final_pending = await agent.load_pending_hitl_request()
+                classification = classify_terminal_status(
+                    final_pending=final_pending,
+                    answered_question_id=None,  # prompt path
+                    saw_hitl_request_event=auto_detach.detached,
+                )
+                if classification.clear_pending:
+                    await cp.save_pending_request(conversation_id, None)
+                final_status = classification.status
```
[the remark this answers is not in the document]
Comment (xfgong) [reply backend/cubebox/streams/run_manager.py:1611]: Fixed in a4fb509a. Backend `_execute_run` now sets `data.paused = true` on `DoneEvent` when `final_status == "paused_hitl"`. Frontend `consumeRunStream` reads `event.data.paused`, sets a new `sawPausedDone` flag instead of `sawDone`, and routes to `finalizePausedStream` (new) which commits the in-flight assistant message + tool results but PRESERVES `pendingConfirmMap` / `pendingAsk`. Card stays visible from the same tab; no reload needed. Both the SSE-direct (`send`) path and the Redis-replay (`consumeRunStream`) path are updated.

**Comment by xfgong** [comment]
@codex please take another pass — pushed a4fb509a addressing both P1s: (1) paused teardown keeps the Redis lock + start_run guards DB pending up-front; (2) DoneEvent carries data.paused=true on pause + frontend finalizePausedStream preserves pending HITL slots.

**Comment by xfgong** [comment]
@codex please do one COMPREHENSIVE pass on the latest commit (2c1bcb35) — surface every P1 / P2 / nit you'd raise across the whole diff in a single review rather than batched. Focus areas: durable HITL pause/respond correctness (CheckpointedChannel, claim_resume CAS, paused_hitl state machine, dangling cleanup, finalize CAS), the integration between main's reflection trigger and my factory extraction (extra_ref_holder ferry pattern), all the error-path edge cases (worker crash mid-pause, two-tab race, long-pause TTL recovery, org policy change mid-pause), and the cubepi v3 schema migration. Don't hold anything back — I'll triage; we'd rather have one full list than incremental rounds.

File: backend/cubebox/streams/run_manager.py [hunk 3]
```diff
@@ -999,927 +1216,1700 @@ async def _run_cubepi_path(
...
-                            _sess,
-                        )
-
-                _builtin_tools.extend(
-                    _action_tools_for_run(
-                        _action_ctx_factory,
-                        allow_mutations=(trigger == "interactive"),
+                if v is not None
+            }
+            final_status: str = "completed"
```
[the remark this answers is not in the document]
Comment (xfgong) [reply backend/cubebox/streams/run_manager.py]: Fixed in fbd1a32b. `_run_cubepi_respond_path` now defaults `final_status = "errored"` (commented inline: matches `_execute_run`'s same fix from a4fb509a, and the spec's "DB pending stays for retry on respond failure" path). Success branch still overwrites via `classify_terminal_status`.

File: backend/cubebox/api/routes/v1/conversations.py [hunk 4]
```diff
@@ -1048,10 +1086,36 @@ async def cancel_active_run(
     active_run = await get_active_run(
         rds.client, prefix=rds.key_prefix, conversation_id=conversation_id
     )
-    if active_run is None or active_run.status != "running":
+    if active_run is None:
         return {"status": "no_active_run", "run_id": None}
```
[the remark this answers is not in the document]
Comment (xfgong) [reply backend/cubebox/api/routes/v1/conversations.py]: Fixed in fbd1a32b. Cancel route reads `load_pending_run_id` as a fallback when `active_run is None`, then dispatches to `cancel_paused_run` with the persisted run_id. Mirrors the same DB fallback that bootstrap + answer routes already use. User can now cancel a HITL card that survived past Redis TTL.

**Comment by xfgong** [comment]
@codex re-tag — fixed all three from your last review in fbd1a32b: (1) respond path defaults to errored; (2) respond DoneEvent stamps data.paused=true on chained HITL; (3) cancel route falls back to load_pending_run_id for TTL-expired paused. Please continue with the comprehensive pass I asked for — surface anything else you'd flag across the whole branch in one go.

File: frontend/packages/web/components/chat/MessageList.tsx [hunk 5]
```diff
@@ -144,30 +195,40 @@ export function MessageList({ conversationId }: MessageListProps) {
...
       if (workspaceId) client.setWorkspaceId(workspaceId)
-      await submitSandboxConfirm(client, convId, pending.question_id, decision)
-      // Optimistic removal — sandbox_confirm_resolved SSE will also clean up
-      useMessageStore.setState((s) => {
-        const next = { ...s.pendingConfirmMap }
-        delete next[toolCallId]
-        return { pendingConfirmMap: next }
-      })
+      try {
+        await submitSandboxConfirm(client, convId, questionId, decision)
```
[the remark this answers is not in the document]
Comment (xfgong) [reply frontend/packages/web/components/chat/MessageList.tsx:202]: Fixed in 80054cc9. Both handleSandboxConfirm and handleAskUserSubmit now await loadMessages(client, convId) after the optimistic clear — re-bootstraps and consumeRunStream re-attaches to the new active_run (same run_id reused on backend). Continuation events flow without manual reload.

File: backend/cubebox/streams/run_manager.py [hunk 6]
```diff
@@ -1001,1060 +1218,1835 @@ async def _run_cubepi_path(
...
+                    ("run_id", run_id),
+                    ("conversation_id", conversation_id),
+                    ("user_id", ctx.user_id),
+                    ("org_id", ctx.org_id),
+                    ("workspace_id", ctx.workspace_id),
+                    ("turn_kind", "respond"),
+                )
+                if v is not None
+            }
+            final_status: str = "completed"
```
[the remark this answers is not in the document]
Comment (xfgong) [reply backend/cubebox/streams/run_manager.py]: Already fixed in fbd1a32b before this comprehensive review landed — _run_cubepi_respond_path now defaults final_status to errored so an exception path doesnt finalize as completed. Apologies for the cross-thread duplicate; codex was reviewing 2c1bcb35 while my fix was in flight.

File: backend/cubebox/streams/run_manager.py [hunk 7]
```diff
@@ -2359,49 +3325,53 @@ async def publish_stream_event(sse_event: AgentEvent, agent_key: str | None) ->
                     },
```
[the remark this answers is not in the document]
Comment (xfgong) [reply backend/cubebox/streams/run_manager.py]: Already fixed in fbd1a32b — _execute_respond_run reads final_meta.status before emitting DoneEvent and stamps data.paused=true when paused_hitl. Same shape as prompt path. Cross-fired with comprehensive review timing.

File: frontend/packages/core/src/stores/messageStore.ts [hunk 8]
```diff
@@ -952,8 +1088,8 @@ export const useMessageStore = create<MessageStore>((set, get) => ({
         pendingSteers: { ...s.pendingSteers, [conversationId]: [] },
         toolStartedMap: {},
         toolResultMap: {},
-        pendingConfirmMap: {},
-        pendingAsk: null,
+        pendingConfirmMap: seedPendingConfirmMap,
+        pendingAsk: seedPendingAsk,
```
[the remark this answers is not in the document]
Comment (xfgong) [reply frontend/packages/core/src/stores/messageStore.ts:1118]: Fixed in 80054cc9. loadMessages now treats bootstrap.pending_hitl.run_id as the paused-on-run marker when bootstrap.active_run is null — sets streamingConversationId / currentRunId from pending_hitl.run_id, attaches consumeRunStream, so the card renders, composer-lock gate works, and a resume from any worker reaches this tab.

File: backend/cubebox/api/routes/v1/conversations.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (xfgong) [reply backend/cubebox/api/routes/v1/conversations.py]: Already fixed in fbd1a32b — cancel route falls back to load_pending_run_id when active_run is None, then dispatches to cancel_paused_run. Same DB fallback shape as the answer routes.

File: frontend/packages/web/components/chat/MessageList.tsx [hunk 9]
```diff
@@ -144,30 +195,40 @@ export function MessageList({ conversationId }: MessageListProps) {
...
+        // Optimistic removal — sandbox_confirm_resolved SSE will also clean up
```
[the remark this answers is not in the document]
Comment (xfgong) [reply frontend/packages/web/components/chat/MessageList.tsx:203]: Already fixed in 80054cc9 (frontend handleAskUserSubmit / handleSandboxConfirm await loadMessages after the optimistic clear → consumeRunStream reattaches to the resumed run). Cross-fired with this review on fbd1a32b.

File: backend/cubebox/streams/run_manager.py [hunk 10]
```diff
@@ -674,76 +796,184 @@ async def dispatch_steer(self, run_id: str, content: str, steer_id: str) -> str:
...
+
+        # CAS-guarded terminal write — a racing flow that took over the
+        # slot wins the row; our finalize is a no-op.
+        await finalize_run_meta_if_claim_matches(
+            self._redis,
+            prefix=self._key_prefix,
+            run_id=run_id,
+            claim_token=claim.claim_token,
+            status="cancelled",
+        )
```
[the remark this answers is not in the document]
Comment (xfgong) [reply backend/cubebox/streams/run_manager.py]: Fixed in b258e6fb. cancel_paused_run now calls clear_active_run + expire_run_data after a winning finalize CAS (mirroring _execute_run cleanup on completed/errored). Test stubs both helpers and asserts they are awaited. Skipped when our claim token lost the CAS — we don't clobber whatever flow took over.

**Comment by xfgong** [comment]
@codex re-tag — b258e6fb addresses the active-run-lock cleanup after paused cancel; 80054cc9 already addressed the reattach-SSE-after-answer P1. Comprehensive pass should be complete now. One more sweep?

File: frontend/packages/core/src/stores/messageStore.ts [hunk 11]
```diff
@@ -784,6 +832,51 @@ async function finalizeCompletedStream(
...
+    messages: {
+      ...state.messages,
+      [conversationId]: [
+        ...(state.messages[conversationId] ?? []),
+        assistantMessage,
+        ...toolMessages,
+      ],
+    },
+    isStreaming: false,
+    streamingConversationId: null,
```
[the remark this answers is not in the document]
Comment (xfgong) [reply frontend/packages/core/src/stores/messageStore.ts]: Fixed in a0c2f381. finalizePausedStream now keeps `streamingConversationId = conversationId` while `pendingAsk` or `pendingConfirmMap` is still attached, so `MessageList` keeps rendering `<AskUserCard>` on both the live-paused path and the bootstrap replay path. `isStreaming` remains false, so the existing `isStreaming && streamingConversationId === ...` guards continue to gate correctly.

**Comment by xfgong** [comment]
@codex pushed a0c2f381 — paused HITL card now stays visible (finalizePausedStream mirrors bootstrap's pending_hitl branch). Any other issues from this round? Please include CRITICAL + HIGH together if so.
